components/ntp_background_images/browser/ntp_background_images_service.cc

```diff
@@ -0,0 +1,498 @@
+
+  is_super_referrer_active_ = true;
+  if (IsAlreadyRegistered(*value->FindStringKey(kComponentID)))
+    return;
+
+  local_pref_->Set(prefs::kNewTabPageCachedSuperReferrerComponentInfo, *value);
+
+  UnRegisterSponsoredImagesComponentIfRunning();
+
+  DVLOG(2) << __func__ << ": Start NTP SR component";
+  DCHECK(IsValidSuperReferrerComponentInfo(*value));
+  RegisterNTPBackgroundImagesComponent(
+      cus_,
+      *value->FindStringKey(kPublicKey),
+      *value->FindStringKey(kComponentID),
+      base::StringPrintf("NTP Super Referrer (%s)",
+                         value->FindStringKey(kCompanyName)->c_str()),
+      base::BindRepeating(&NTPBackgroundImagesService::OnComponentReady,
+                          weak_factory_.GetWeakPtr(),
+                          true));
+}
+
+void NTPBackgroundImagesService::StartSponsoredImagesComponent() {
+  const std::string locale =
+      brave_ads::LocaleHelper::GetInstance()->GetLocale();
+  const auto& data = GetSponsoredImagesComponentData(
+      brave_ads::LocaleHelper::GetCountryCode(locale));
+  if (!data)
+    return;
+
+  is_super_referrer_active_ = false;
+  if (IsAlreadyRegistered(data->component_id))
+    return;
+
+  // If SI is not running, unregister if SR is running now.
+  UnRegisterSuperReferrerComponentIfRunning(GetCachedReferralPromoCode());
+
+  DVLOG(2) << __func__ << ": Start NTP SI component";
+  RegisterNTPBackgroundImagesComponent(
+      cus_,
+      data->component_base64_public_key,
+      data->component_id,
+      base::StringPrintf("NTP Sponsored Images (%s)", data->region.c_str()),
+      base::BindRepeating(&NTPBackgroundImagesService::OnComponentReady,
+                          weak_factory_.GetWeakPtr(),
+                          false));
+}
+
+void NTPBackgroundImagesService::UnRegisterSuperReferrerComponentIfRunning(
+    const std::string& referral_code) {
+  if (!mapping_table_value_.is_dict())
+    return;
+
+  if (!cus_)
+    return;
+
+  if (const base::Value* value =
+          mapping_table_value_.FindDictKey(referral_code))
+    cus_->UnregisterComponent(*value->FindStringKey(kComponentID));
+}
+
+void NTPBackgroundImagesService::OnGetMappingTableData(
+    std::unique_ptr<std::string> json_string) {
+  if (!json_string) {
+    DVLOG(2) << __func__ << ": Failed to fetch mapping table."
+             << " Start NTP SR or SI.";
+    const auto* value = local_pref_->Get(
+        prefs::kNewTabPageCachedSuperReferrerComponentInfo);
+    if (IsValidSuperReferrerComponentInfo(*value)) {
+      // Generate temporal mapping table and start SR.
+      const std::string cached_promo_code = GetCachedReferralPromoCode();
+      mapping_table_value_ = base::Value(base::Value::Type::DICTIONARY);
+      mapping_table_value_.SetKey(cached_promo_code, value->Clone());
+      StartSuperReferrerComponent(cached_promo_code);
+      return;
+    }
+    // In network error, start SR if we have cached info. If not, start SI.
+    // TODO(simonhong): Should we retry?
```

Comment:
Retry will be done when browser launches again, right? This might be sufficient. When going over the spec, we did discuss trying again for a 24 hour period